operators/tokenizer/bpe_utils.hpp

```diff
@@ -99,189 +93,169 @@ class SpecialTokenMap {
-
-  void Set(std::u32string_view val) {
-    m_text = val;
-  }
-
-  std::pair<bool, std::u32string> GetNextToken(std::string & regex_expr) {
-    while (!m_text.empty()) {
-      auto res = RegexMatchCustom(regex_expr);
-      if (res.empty()) {
-        m_text = m_text.substr(1);
-        continue;
-      }
-      return {true, res};
-    }
-    return {false, {}};
-  }
-
- public:
-
-  // Although we have RegexMatchGeneral which performs regex matching given any general regex string
-  // (not just GPT2, Llama, or C++ standard library compatible regex strings), we write the following
-  // two methods for GPT2 and Llama in order to make performance improvements.
-
-  std::u32string RegexMatchGPT2() {
+  static constexpr const char GPT2_REGEX_PATTERN[] =
+      R"('s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+)";
+  static constexpr const char LLAMA_REGEX_PATTERN[] =
+      R"((?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+)";
 
-    // Case 1: English apostrophe handling (1st-7th Alternative, the endings for her's, can't, you're, etc.)
+  PreTokenizerWithRegEx() = default;
 
-    // Note: the sequencial of the following if should not be switched, which follows the python regex's syntax
-
-    // Standard Library Search might be too compute intensive here due to conversions to and fro ustring and wstring,
-    // so we stick to manual parsing here for efficiency, however (as an example for the usage of RegexMatchSTD),
-    // note that this following snippet would also work:
-
-    // std::u32string_view std_regex = RegexMatchSTD(U"'s|'t|'re|'ve|'m|'ll|'d");
-    // if (std_regex.size() != 0){
-    //   return std_regex;
-    // }
-
-    // 's|'t|'re|'ve|'m|'ll|'d
+  // "'s|'t|'re|'ve|'m|'ll|'d"
+  std::u32string_view Match_GPT2_Pattern_1() {
     if ((m_text[0] == U'\'') && (m_text.size() > 1)) {
-      if ((m_text[1] == U's') || (m_text[1] == U't') ||
-          (m_text[1] == U'm') || (m_text[1] == U'd')) {
-        std::u32string res = m_text.substr(0, 2);
+      if ((m_text[1] == U's') || (m_text[1] == U't') || (m_text[1] == U'm') || (m_text[1] == U'd')) {
+        std::u32string_view res = m_text.substr(0, 2);
         m_text = m_text.substr(2);
         return res;
       }
 
       if (m_text.size() > 2) {
-        if (((m_text[1] == U'r') && (m_text[2] == U'e')) ||
-            ((m_text[1] == U'v') && (m_text[2] == U'e')) ||
+        if (((m_text[1] == U'r') && (m_text[2] == U'e')) || ((m_text[1] == U'v') && (m_text[2] == U'e')) ||
             ((m_text[1] == U'l') && (m_text[2] == U'l'))) {
-          std::u32string res = m_text.substr(0, 3);
+          std::u32string_view res = m_text.substr(0, 3);
           m_text = m_text.substr(3);
           return res;
         }
       }
     }
 
-    // Case 2: Language Character Class (8th Alternative)
```

Comment:
can we keep comments explaining different regex cases for better understanding when looking at/editing this code in the future?

Reply:
This information can easily be found https://regex101.com/ and they are standard regex pattern, so I prefer not to elaborate them to every functions